botorch/optim/optimize.py

```diff
@@ -114,6 +114,19 @@ def __post_init__(self) -> None:
                     f"shape is {batch_initial_conditions_shape}."
                 )
 
+            if (
+                self.raw_samples is not None
+                and (self.raw_samples - batch_initial_conditions_shape[-2]) > 0
+                and len(batch_initial_conditions_shape) == 3
+                and self.num_restarts is not None
+                and self.num_restarts != batch_initial_conditions_shape[0]
+            ):
+                raise ValueError(
+                    "If using `batch_initial_conditions` together with `raw_samples`, "
+                    "the first repeat dimension of `batch_initial_conditions` must "
+                    "match `num_restarts`."
+                )
+
```

Comment:
@saitcakmak most of the tests that break appear to do so based on this setting in which we see that if we give a 3 dimension initial condition tensor if the first dimension isn't equal to num_restarts we can't concat. Not sure the cleanest way to handle this edge case, in intentional/correct use of the code it shouldn't occur but  it may create issues depending on how people have set these arguements.

Reply:
So you mean in existing use cases in which previously folks were providing both `batch_initial_conditions` and `raw_samples` but only the `batch_initial_conditions` were used? I think it's probably fine to error out like this. An alternative would be to fall back to the previous behavior of ignoring `raw_samples` in this case but raising some `DeprecationWarning` that this behavior will change going forward. Then you can both achieve backward compatibility and allow for your use case (for some slight increase in complexity, at least temporarily).

Reply:
previous the `n_restarts` kwarg was just unused if giving initial batch conditions. As far as I understand the 3D are `n_restarts, n_samples, n_dims`. If we are giving initial conditions we can within the code choose to give those conditions either as `n_samples, 1, n_dims`, `1, n_samples, n_dims`, I even think it wouldn't matter if they were given as `2, n_samples/2, n_dims`, or something like that. 

If n_samples-n_ic_samples is bigger than 1 we would expect to see different solutions from the optimization and so it's useful to have repeats. However, the points from the IC generator will always be returned as `n_restarts, n_samples-n_ic_samples, n_dims` so if the batch initial conditions are not size 1 (so they can be easily repeated) or already equal to `n_restarts` in the 0th dimension then it makes it fairly messy, in principal you could do some sort of repeat and interleaving to get the combinations right but then the number of repeats considered wouldn't be `n_repeats`.

Reply:
In terms of the solution I am happy to either add the deprecation warning or leave as is, whatever makes more sense to you.